Core: don't cache the get_all_state result

alwaysintreble opened the pull request:
## What is this fixing or adding?
This removes the caching of all_state completely. This method is typically used for world pre_fill stages and as such is inaccurate, since the items being placed in that step are typically non-existent in the state. This means worlds attempting to use the cached state end up with bugs, and it can be incorrect at any point as there's no guarantee on how early or late the cached state was first created.

## How was this tested?
Just tests.

qwint commented on BaseClasses.py:
```diff
@@ -427,11 +427,18 @@ def get_entrance(self, entrance_name: str, player: int) -> Entrance:
-            return cached.copy()
+    def get_all_state(self, use_cache: bool = False, allow_partial_entrances: bool = False) -> CollectionState:
+        """
+        Creates a new CollectionState, and collects all pre_fill items, all items in the multiworld itempool, those
+        specified in each worlds' `get_pre_fill_items()`, and then sweeps the multiworld collecting any other items
```
"collects all pre_fill items . . . [and] those specified in each worlds' `get_pre_fill_items()`"
is this not duplicative? if it's somehow about two different things you should be more verbose so it's obvious

qwint commented on BaseClasses.py:
```diff
@@ -427,11 +427,18 @@ def get_entrance(self, entrance_name: str, player: int) -> Entrance:
+        it is able to reach, building as complete of a completed game state as possible.
+
+        :param use_cache: Deprecated and unused.
+        :param allow_partial_entrances: Whether the CollectionState should allow for disconnected entrances while
+         sweeping, such as during entrance randomization.
```
this should probably be "such as before entrance randomization is complete" or something similar, as worlds without ER using get all state before the connect_entrances should be using the arg

qwint reviewed (approved):
updated docs look good, code changes are minimal and look fine assuming we want to remove the caching functionality (which i'm personally on the side of as it seems very unsafe to actually keep track of if the state is actually current and correct, I'm also trusting treble's other messages mentioning that using cache is usually slower anyways)

pure code review; no tests/gens/etc.

alwaysintreble commented:
> I'm also trusting treble's other messages mentioning that using cache is usually slower anyways

yeah I didn't specify that here because I don't have hard proof of that, but I *have* seen it. I have no idea how creating a copy could be slower than rebuilding the entire state, but it probably depends heavily on plenty of factors, and I didn't test against that here so it's a little bit viable it's no longer the case since we have changed some of `CollectionState`'s attributes and I think we've changed how a few of the copies are made, since that anecdotal evidence.

NewSoupVi commented:
What do you think about a deprecation message like this?

```def get_all_state(self, use_cache: bool | None = None, allow_partial_entrances: bool = False) -> CollectionState:
    if use_cache is not None:
        Utils.deprecate(...)
```

or probably a warning for now, and then Utils.deprecate later.

alwaysintreble commented:
> What do you think about a deprecation message like this?

I considered this but the reason I didn't is because there are over 20 unique calls of this function and add on to that multiple players for a single world using this... Still probably a good idea though so maybe lock it behind debug for now with a TODO?

Exempt-Medic reviewed (approved):
I'm approving this, but think other core devs need to talk about it / rule on it / merge it

NewSoupVi reviewed (approved):
I agree with this
